crates/red_knot/src/types/eval.rs

```diff
@@ -0,0 +1,141 @@
+#![allow(dead_code)]
+use crate::db::{HasJar, SemanticDb, SemanticJar};
+use crate::module::ModuleName;
+use crate::symbols::{Definition, ImportFromDefinition, SymbolId};
+use crate::types::Type;
+use crate::FileId;
+use ruff_python_ast::{AstNode, StmtClassDef};
+
+#[tracing::instrument(level = "trace", skip(db))]
+pub fn eval_symbol<Db>(db: &mut Db, file_id: FileId, symbol_id: SymbolId) -> Type
+where
+    Db: SemanticDb + HasJar<SemanticJar>,
+{
+    let symbols = db.symbol_table(file_id);
+    let defs = symbols.defs(symbol_id);
+
+    if let Some(ty) = db
+        .jar()
+        .type_store
+        .get_cached_symbol_type(file_id, symbol_id)
+    {
+        return ty;
+    }
+
+    // TODO handle multiple defs, conditional defs...
+    assert_eq!(defs.len(), 1);
+
+    let ty = match &defs[0] {
+        Definition::ImportFrom(ImportFromDefinition {
+            module,
+            name,
+            level,
+        }) => {
+            // TODO relative imports
+            assert!(matches!(level, Some(0)));
+            let module_name = ModuleName::new(module.as_ref().expect("TODO relative imports"));
+            if let Some(module) = db.resolve_module(module_name) {
+                let remote_file_id = module.path(db).file();
+                let remote_symbols = db.symbol_table(remote_file_id);
+                if let Some(remote_symbol_id) = remote_symbols.root_symbol_id_by_name(name) {
+                    db.eval_symbol(remote_file_id, remote_symbol_id)
+                } else {
+                    Type::Unknown
+                }
+            } else {
+                Type::Unknown
+            }
+        }
+        Definition::ClassDef(node_key) => {
+            let node_key = node_key.erased();
+            if let Some(ty) = db.jar().type_store.get_cached_node_type(file_id, node_key) {
+                ty
+            } else {
+                let parsed = db.parse(file_id);
+                let ast = parsed.ast();
+                let node = StmtClassDef::cast_ref(
+                    node_key
+                        .resolve(ast.as_any_node_ref())
+                        .expect("node key should resolve"),
+                )
+                .expect("node key should cast");
+
+                let store = &mut db.jar_mut().type_store;
+                let ty = store.add_class(file_id, &node.name.id);
+                store.cache_node_type(file_id, *node_key, ty);
+                ty
+            }
```

Comment:
Nit
```suggestion
            let Some(ty) = db.jar().type_store.get_cached_node_type(file_id, node_key) else {
                let parsed = db.parse(file_id);
                let ast = parsed.ast();
                let node = StmtClassDef::cast_ref(
                    node_key
                        .resolve(ast.as_any_node_ref())
                        .expect("node key should resolve"),
                )
                .expect("node key should cast");

                let store = &mut db.jar_mut().type_store;
                let ty = store.add_class(file_id, &node.name.id);
                store.cache_node_type(file_id, *node_key, ty);
                ty
            }
```

Reply:
The suggested change here doesn't work as-is, because this arm of the match needs to actually evaluate to a Type (ty). From what I can tell, it seems like `let-else` is for cases where the `else` diverges, but doesn't evaluate to an expression. I could restructure this to use a mutable `ty` I guess, but that doesn't seem like an improvement over the `if-let-else` I have now.

Let me know if I'm missing a way for this suggestion to work out better than the `if-let-else`!